### ISSUE TITLE
SegFault created via BPF

### ISSUE THREAD

**Issue by computer5student** [iovisor/bcc#1470]
Hello,

I am not sure whether this is a problem with bcc, so please forgive me if it isn't, however, I think the behavior of the program I created is very strange, so I will post it here.

My intention was to create a BPF hashmap whose keys are character arrays of fixed size and whose values are u32. Length of the character array is specified as user input before the BPF program is attached to any hook.
So I defined the following map:

```
// DROPMAP_SIZE_KEY is given to the BPF program via cflags
typedef char Key[DROPMAP_SIZE_KEY];
BPF_HASH(dropmap, Key, u32, 30);
```

Now I want to only use the first key-value pair of this map to count packet drops in XDP:

```
int xdpmdrp(struct xdp_md *ctx) {
         u32 *value_p;
         u32 zero=0;
         Key key = {0};
         value_p = dropmap.lookup_or_init(&key, &zero);

         void* data_end = (void*)(long)ctx->data_end;
         void* idata = (void*)(long)ctx->data;
         struct ethhdr *eth = idata;
         struct iphdr *iph = (struct iphdr*) (eth+1);
         struct udphdr *udph = (struct udphdr*) (iph+1);

         if(udph + 1 > data_end) {
                 return XDP_PASS;
         }
         if(value_p) *value_p+=1;
                  return XDP_DROP;
```

From a python userspace program I attach this program to the XDP hook and packet counting begins. If I do this with a 32-bit sized key everything works fine and I can see the counters going up. The output of my python program looks the following (I sent 1000 packets while I was measuring):

```
Content of the map: c_char('\x00') 1000
Entries in dropmap:  1
Removing XDP filter from device...
Finished
```

Now I do everything the same but with a key length of 64 byte (or anything larger than 32) and the output of my python program is the following:

```
Content of the map: c_char('\x00') 1000
Entries in dropmap:  2
Segmentation fault (core dumped)
```

I have already troubleshooted that the segfault occurs at the following place:

```
for k,v in dropmap.items():
     do_something()
```

The first iteration of the loop works normally, the second creates the segfault, probably because I only inserted one entry into the hashmap but len(dropmap) returns 2.

Is this happening because of a programing mistake of mine or have I encountered a bug?  If it is my mistake, shouldn't the verifier warn me somehow or better block my program from entering the kernel?

I already upgraded bcc via apt, but nothing changed. I am running Ubuntu 17.04 with Kernel 4.14.0-rc4+.

Kind regards,
Computer Student

**Comment by yonghong-song** [comment]
Could you post your whole program here? This will make debugging easier. I suspect that this has something to do with key type as an array. But it would be good if you post a reproducible test case here.

**Comment by computer5student** [comment]
I think everything interesting is already listed above. My python program is quite large, but I will try nonetheless to give the parts where the map is involved:

BPF Kernelspace Program:

```
#include <uapi/linux/ptrace.h>
#include <linux/blkdev.h>
#include <uapi/linux/bpf.h>
#include <linux/in.h>
#include <linux/if_ether.h>
#include <linux/if_packet.h>
#include <linux/ip.h>
#include <uapi/linux/udp.h>
#include <net/sock.h>
#include <bcc/proto.h>

typedef char Key[DROPMAP_SIZE_KEY];
BPF_HASH(delmap, u32, struct deldata, 300000);
BPF_HASH(dropmap, Key, u32, 30);

static inline bool relevant_packet_udp_socket(u16 h_proto, u8 protocol, u16 dest_port, u32 dest_ip) {
	return ((h_proto == ETH_P_IP) & (protocol == 17) & (dest_port == DEST_PORT) & (dest_ip == DEST_IP));
}

int xdpmdrp(struct xdp_md *ctx) {
	u32 *value_p;
	u32 zero=0;
        // interpreted as a char-array this value is 0000
	Key key = {204484656};
	value_p = dropmap.lookup_or_init(&key, &zero);

	void* data_end = (void*)(long)ctx->data_end;
        void* idata = (void*)(long)ctx->data;
        struct ethhdr *eth = idata;
	struct iphdr *iph = (struct iphdr*) (eth+1);
	struct udphdr *udph = (struct udphdr*) (iph+1);

	if(udph + 1 > data_end) {
		return XDP_PASS;
	}

	u16 dest_port = ntohs(udph->dest);
	u32 dest_ip = ntohl(iph->daddr);
	u8 protocol = iph->protocol;
	u16 h_proto = ntohs(eth->h_proto);

	if(relevant_packet_udp_socket(h_proto, protocol, dest_port, dest_ip)) {
		if(value_p) *value_p+=1;
		return XDP_DROP;
	}
	// Allow ARP
	if(h_proto == ETH_P_ARP)
		return XDP_PASS;
	return XDP_DROP;
}
```

Relevant parts of Python Userspace Program:

```
#!/usr/bin/env python
from __future__ import print_function
from bcc import BPF
import ctypes as ct
import pyroute2
import time
import sys
import socket
import struct
import os
import subprocess
from datetime import datetime
from pyroute2 import IPRoute, protocols
import numpy as np
import matplotlib as mpl
mpl.use('agg')
from threading import Thread

#Defines
XDP = 0

# Cleans up all the hooked programs
def removeBPF(bpf, iproute, ifindex, error=None):
    if error:
        print("\nERROR occured: {}".format(error))
        print("Cleaning up and shutting down the program...")
    print("Removing XDP filter from device...")
    bpf.remove_xdp(device, flags)
    if iproute != None:
        print("Removing tc filter from device...")
        #if "ifindex" in locals(): ip.link_remove(ifindex)
        iproute.tc("del", "ingress", ifindex, "ffff:")
    print("Finished")
    sys.exit(0)

# Print measured information
def print_dropdata(dropmap):
    print('Entries in Dropmap {}:'.format(len(dropmap)))
    for k,v in dropmap.iteritems():
        print('Entries in Dropmap {} {}'.format(k, v.value))

flags = 0
device = eth0
bip = 10.0.0.100
udp_port = 1337
hook = BPF.XDP
bpf_file = "bpf/bcc_bpf.c"
dropmap_size_key = 64
    try:
        ##  Prepare and install the BPF program
        b = BPF(src_file=bpf_file, cflags=["-w",
            "-DDROPMAP_SIZE_KEY=%d" % dropmap_size_key,
            "-DCTX_TYPE=xdp_md",
            "-DDEST_PORT=%d" % udp_port,
            "-DDEST_IP=%d" % ipv4_strToInt(bip)])
        fn = b.load_func("xdpmdrp", hook)
        if(hook == BPF.XDP):
            b.attach_xdp(device, fn, flags)

        # Prepare BPF maps
        dropmap = b.get_table("dropmap")
        dropmap.clear()

        print("Taking measurements... Press CTRL+C to stop")
        # Measure time for 10 seconds
        thread_timer = Thread(target=time.sleep, args = (10, ))
        thread_timer.start()

        while (thread_timer.isAlive()):
            #### Do this stuff while BPF and userspace programs capture data
            time.sleep(4)

        # Here the error occurs
        print_dropdata(dropmap)


    except KeyboardInterrupt:
         removeBPF(b,None, None, "Aborted by User")


```

As described the problem seems to be the following: The BPF program only uses one fixed field to write into, namely the char-array of 0000, so only one field of the BPF_HASHMAP is used. However, the Python program gets the information that 2 entries in the map exist and the part

```
    for k,v in dropmap.iteritems():
        print('Entries in Dropmap {} {}'.format(k, v.value))
```

then triggers an error in the 2nd iteration, because it somehow gets into trouble accessing this non-existent, second entry.

Again, output of my program for key-length of 64 bytes:

```
Taking measurements... Press CTRL+C to stop
Entries in Dropmap 2:
Content of the map: c_char('0') 1000
Segmentation fault (core dumped)
```

**Comment by yonghong-song** [comment]
I am actually confused with the code:
```
        typedef char Key[DROPMAP_SIZE_KEY];
...
	Key key = {204484656};
	value_p = dropmap.lookup_or_init(&key, &zero);
...
```
I thought Key is an array type. So if you do `Key key = {204484656}`, you try to do `key[0] = 204484656` which of course overflows. And `&key` actually pass the pointer to the array pointer?

Did I miss anything here?

**Comment by computer5student** [comment]
Hi,

my intention was that I only use one key-value pair to store the number of packet drops. So I have one key which is a 8-byte char array and contains only 0s. So what I did was to look up the ASCII table to see that 0b 0011 0000 would represent a character 0 and I concatenated this and got 204484656. Since the array as a whole is only a single key, I thought that 

`Key key = {204484656};`

would be fine and it also worked.

`value_p = dropmap.lookup_or_init(&key, &zero);`

Here I just took the syntax that I found in the bcc reference guide:

`Syntax: *val map.lookup_or_init(&key, &zero)`

I use char-arrays as keys, so I give a pointer to a char-array. In the beginning I used it without the & before key, but then map.lookup_or_init always returned zero.


I tested my program with changing the line

Key key = {204484656};

to

**Comment by computer5student** [comment]
Hi,

my intention was that I only use one key-value pair to store the number of packet drops. So I have one key which is a 8-byte char array and contains only 0s. So what I did was to look up the ASCII table to see that 0b 0011 0000 would represent a character 0 and I concatenated this and got 204484656. Since the array as a whole is only a single key, I thought that

`Key key = {204484656};`

would be fine and it also worked.

`value_p = dropmap.lookup_or_init(&key, &zero);`

Here I just took the syntax that I found in the bcc reference guide:

Syntax: *val map.lookup_or_init(&key, &zero)

I use char-arrays as keys, so I give a pointer to a char-array. In the beginning I used it without the & before key, but then map.lookup_or_init always returned zero.

However, I tried out your ideas and I see that if I change 

`Key key = {204484656};`
to 
`Key key = {48};`

and try out the program again I get the verifier error that I expected:

5: (7b) *(u64 *)(r10 -520) = r6
invalid stack off=-520 size=8

HINT: Looks like you exceeded the BPF stack limit. This can happen if you allocate too much local variable storage. For example, if you allocated a 1 Kbyte struct (maybe for BPF_PERF_OUTPUT), busting a max stack of 512 bytes.


Nonetheless I wonder of it is a desired behavior of the verifier to allow me an overflow like this?

**Comment by computer5student** [comment]
Update:

I fixed my program according to the feedback here and the BPF part looks now like this:

```
typedef u8 Key[DROPMAP_SIZE_KEY];

int xdpmdrp(struct xdp_md *ctx) {                                                                                                                                                                                     
    u32 *value_p;
    u32 zero=0;

    Key key = {};
    value_p = dropmap.lookup_or_init(key, &zero);

    if(value_p) value_p+=1;
```

When I increase the size to 64 the problem with the 2 entries is fixed, the map now only contains 1 entry but the segmentation fault still occurs:


Entries in Dropmap 1:
Content of the map: c_ubyte(0) 1000
Segmentation fault (core dumped)

**Comment by computer5student** [comment]
I will close this topic since I have solved the problem. I used the BCC Reference guide all the time as a reference to the bcc python functions and there the get_table() function syntax was given as

`BPF.get_table(name)`

However, I realized that when using a byte-array as I did in this case, this usage does not work because Python gets somehow confused with the keytype and uses ubyte instead if ubyte array. After having a look at the implementation of the get_table() function I realized that I can explicitly tell Python which keytype to use:

`def get_table(self, name, keytype=None, leaftype=None, reducer=None):`

This solves the problem and the segfault of the userspace program does not occur anymore. Given that the BPF verifier should only protect from problems / crashes in the kernel the behavior of letting a userspace program crash does not seem that critical.
